clang/include/clang/3C/OptionalParams.h

```diff
@@ -0,0 +1,213 @@
+// (https://en.cppreference.com/w/cpp/preprocessor/replace).
+#define _OP_COUNT_ARGS(...)                                                    \
+  _OP_MSVC_VA_ARGS_WORKAROUND(                                                 \
+      _OP_GET_ARG7(__VA_ARGS__, 6, 5, 4, 3, 2, 1, _dummy))
+
+// Each _i argument is expected to be of the form `field = expr`, so we generate
+// a statement of the form `_s.field = expr;` to set the struct field.
+#define _OP_ASSIGN_FIELDS_1(_i1) _s._i1;
+#define _OP_ASSIGN_FIELDS_2(_i1, _i2)                                          \
+  _s._i1;                                                                      \
+  _s._i2;
+#define _OP_ASSIGN_FIELDS_3(_i1, _i2, _i3)                                     \
+  _s._i1;                                                                      \
+  _s._i2;                                                                      \
+  _s._i3;
+#define _OP_ASSIGN_FIELDS_4(_i1, _i2, _i3, _i4)                                \
+  _s._i1;                                                                      \
+  _s._i2;                                                                      \
+  _s._i3;                                                                      \
+  _s._i4;
+#define _OP_ASSIGN_FIELDS_5(_i1, _i2, _i3, _i4, _i5)                           \
+  _s._i1;                                                                      \
+  _s._i2;                                                                      \
+  _s._i3;                                                                      \
+  _s._i4;                                                                      \
+  _s._i5;
+#define _OP_ASSIGN_FIELDS_6(_i1, _i2, _i3, _i4, _i5, _i6)                      \
+  _s._i1;                                                                      \
+  _s._i2;                                                                      \
+  _s._i3;                                                                      \
+  _s._i4;                                                                      \
+  _s._i5;                                                                      \
+  _s._i6;
+#define _OP_ASSIGN_FIELDS(_count, ...)                                         \
+  _OP_MSVC_VA_ARGS_WORKAROUND(_OP_ASSIGN_FIELDS_##_count(__VA_ARGS__))
+// In _OP_ASSIGN_FIELDS, ## takes precedence over the expansion of the _count
+// argument, so if we let PACK_OPTS call `_OP_ASSIGN_FIELDS(_OP_COUNT_ARGS(foo),
+// bar)` directly, we'd end up with `_OP_ASSIGN_FIELDS__COUNT_ARGS(foo)(bar)`.
+// _OP_ASSIGN_FIELDS_WRAP expands its _count argument before calling
+// _OP_ASSIGN_FIELDS, working around the problem. See the second example on
+// https://gcc.gnu.org/onlinedocs/cpp/Argument-Prescan.html.
+#define _OP_ASSIGN_FIELDS_WRAP(_count, ...)                                    \
+  _OP_ASSIGN_FIELDS(_count, __VA_ARGS__)
+
+// PACK_OPTS currently doesn't support passing an empty list of fields to the
```

Comment:
Is it easy to extent `PACK_OPTS` to support empty lists of fields? It might look more stylistically consistent this way?

Reply:
> Is it easy to extent `PACK_OPTS` to support empty lists of fields?

`PACK_OPTS`, yes; in fact, I did it in [a previous version of this PR](https://github.com/correctcomputation/checkedc-clang/commit/a9ff8bdda29c4ebc56583f34057c8000e37fbdee), but I removed the code because I didn't think the team would want to be responsible for maintaining it given that we don't need it.  For the function-specific wrapper macros that would be used at call sites, such as `MKSTRING_OPTS`, I think we are out of luck because [passing zero arguments to a variadic macro parameter isn't allowed until C++20](https://en.cppreference.com/w/cpp/preprocessor/replace#Function-like_macros).  We might be able to find a way to make it work with the current build flags on all our actual target compilers, but I wouldn't want to do that work without a good reason.

> It might look more stylistically consistent this way?

What are you saying would look more stylistically consistent?  Call sites?  I think I prefer `mkString(Info.getConstraints())` to `mkString(Info.getConstraints(), MKSTRING_OPTS())`.

Reply:
Ok. Sounds good